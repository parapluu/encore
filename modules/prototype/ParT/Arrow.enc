module Arrow(
    arr,
    first,
    second,
    compose,
    fanout,
    combine,
    liftA2,
    runArrow,
    runParArrow
  )

import ParT

--
-- This module provides functions for creating and operating on a parallel arrows,
-- a.k.a. PArrow. A PArrow can be seen as a function lifted to a ParT
-- (parallel computation). Multiple functions can be lifted to PArrow.
--
-- Use the `arr` combinator to lift a function into a PArrow. This does not
-- execute the function but allows you to declarative compose functions.
--
-- To run a parallel arrow over a single item, use the `runArrow` function.
--
-- To run a parallel arrow over a collection of parallel items, use `runParArrow`.
--
-- Complex combinations are created with the following combinators:
-- * first
-- * second
-- * compose
-- * fanout
-- * combine
-- * liftA2
--


typedef PArrow[a, b] = Par [a -> b]

--
-- arr :: (b -> c) -> PArrow[b, c]
--
-- `arr` lifts a function to the parallel arrow. The PArrow type represents
-- an abstraction with possibly many functions. By lifting multiple
-- functions these will potentially start performing work in parallel.
--
fun arr[a, b](fn : a -> b) : PArrow[a, b]
  liftv(fn)
end

--
-- first :: PArrow[a, b] -> PArrow [a, c] [b, c]
--
-- Given a PArrow that accepts one argument of type `a`
-- and returns an argument of type `b`, return a parallel arrow
-- that takes a tuple `(a, c)` and returns a tuple `(b, c)`.
-- (the second argument in the tuple is not changed).
--
fun first[a, b, c](p : PArrow[a, b]) : PArrow[(a, c), (b, c)]
  val closure = fun (fn : a -> b)
                  liftv(fun (t : (a, c)) => (fn(t.0), t.1))
                end
  bind(closure, p)
end

--
-- second :: PArrow[a, b] -> PArrow[(c, a), (c, b)]
--
-- Given a ParT function that accepts one argument of type `a`
-- and returns an argument of type `b`, return a lifted function
-- that takes a tuple `(c, a)` and returns a tuple `(b, c)`.
-- (the first argument in the tuple is not changed).
--
fun second[a, b, c](p : PArrow[a, b]) : PArrow[(c, a), (c, b)]
  val closure = fun (fn : a -> b)
                  liftv(fun (t : (c, a)) => (t.0, fn(t.1)))
                end
  bind(closure, p)
end

--
-- compose :: PArrow[a, d] -> PArrow[d, c]) -> PArrow[a, c]
--
-- Parallel arrow composition. The output of the first arrow is given as
-- input to the second arrow.
--
fun compose[a, d, c](p : PArrow[a, d], p2 : PArrow[d, c]) : PArrow[a, c]
  bind(fun (fn : a -> d)
         bind(fun (fn2 : d -> c) => arr(fun (x : a) => fn2(fn(x))), p2)
       end, p)
end

--
-- fanout :: PArrow[a, b] -> PArrow[a, c]) -> PArrow[a, (b, c)]
--
-- Given two parallel arrows that take the same first argument,
-- return a new parallel arrow that executes both functions,
-- yielding a tuple `(b, c)` as result.
--
fun fanout[a, b, c](p : PArrow[a, b], p2 : PArrow[a, c]) : PArrow[a, (b, c)]
  val clos = fun (fn1 : a -> b)
               val clos2 = fun (fn2 : a -> c)
                             liftv(fun (x : a) => (fn1(x), fn2(x)))
                           end
               bind(clos2, p2)
             end
  bind(clos, p)
end

--
-- combine :: PArrow[a, b] -> PArrow[c, d]) -> PArrow[(a, c), (b, d)]
--
-- Given two parallel arrows, return a new parallel arrow that takes tuples
-- `(a, c)` and return tuples `(b, d)`.
--
fun combine[a, b, c, d](p1 : PArrow[a, b], p2 : PArrow[c, d]) : PArrow[(a, c), (b, d)]
  val clos = fun (fn1 : a -> b)
               val clos2 = fun (fn2 : c -> d)
                             liftv(fun (t : (a, c)) => (fn1(t.0), fn2(t.1)))
                           end
               bind(clos2, p2)
             end
  bind(clos, p1)
end

--
-- liftA2 :: ((b, c) -> d) -> PArrow[a, b] -> PArrow[a, c] -> PArrow [a, d]
--
-- Given two parallel arrows that take the same argument, compose them and
-- aggregate their results with the function given as first argument.
-- (this resembles a ForkJoin in the sense that two functions may run possibly
--  in parallel and their end result will be aggregated using the function `fn`).
--
fun liftA2[a, b, c, d](fn: (b, c) -> d, p1: PArrow[a, b], p2: PArrow[a, c]): PArrow [a, d]
  val pb = fanout(p1, p2)
  val fnArr = arr(fun (t: (b, c)) => fn(t.0, t.1))
  compose(pb, fnArr)
end

--
-- runArrow :: a -> PArrow[a, b]) -> Par[b]
--
-- Lift the item to the parallel arrow. All functions in the PArrow are applied
-- to the item. return a ParT collection.
--
fun runArrow[a, b](item : a, p : PArrow[a, b]) : Par[b]
  p >> (fun (fn : a -> b) => fn(item))
end

--
-- runParrArrow :: Par[a] -> PArrow[a, b] -> Par[b]
--
-- Given a ParT collection and a parallel arrow, lift each item in the ParT
-- to the arrow function. return a ParT collection.
--
fun runParArrow[a, b](p: Par[a], arrow: PArrow[a, b]): Par[b]
  join(p >> fun (item: a) => runArrow(item, arrow))
end
