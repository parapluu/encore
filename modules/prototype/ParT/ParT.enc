module ParT(
    empty,
    filter,
    bind,
    foreachp,
    aggregate,
    intersection,
    distinct,
    union,
    app,
    pairWith,
    groupJoin,
  )

fun empty[a]() : Par[a]
  EMBED (Par[a])
    new_par_empty(_ctx, runtimeType[0]);
  END
end

fun filter[a](fn : a -> bool, ps : Par[a]) : Par[a]
  join(ps >> fun (item : a) => if fn(item) then
                                 liftv(item)
                               else
                                 empty[a]()
                               end)
end

fun bind[a, b](fn : a -> Par[b], p : Par[a]) : Par[b]
  join(p >> fn)
end

fun foreachp[t](p : Par[t], fn : t -> void) : Par[void]
  p >> fn
end

fun aggregate[t, a, r](p : Par[t], var init : a, fn1 : (a, t) -> a, fn2 : a -> r) : Par[r]
  liftf(reduce(fn1, init, p)) >> fn2
end

fun intersection[t](p1 : Par[t], p2 : Par[t], cmp : (t, t) -> int) : Par[t]
  EMBED (Par[t])
    party_intersection(_ctx, #{p1}, #{p2}, #{cmp}, _enc__type_t);
  END
end

fun distinct[t](p : Par[t], cmp : (t, t) -> int) : Par[t]
  EMBED (Par[t])
    party_distinct(_ctx, #{p}, #{cmp}, _enc__type_t);
  END
end

fun union[t](pl : Par[t], pr : Par[t], cmp : (t, t) -> int) : Par[t]
  distinct(pl ||| pr, cmp)
end

fun app[a, b](p : Par[a -> b], item : a) : Par[b]
  p >> (fun (fn : a -> b) => fn(item))
end

fun pairWith[t, k](p : Par[t], fn : t -> k) : Par[(k, t)]
  p >> (fun (x : t) => (fn(x), x))
end

-- private function.
-- filter out items in the `p` ParT whose key is different from the given key `x`.
-- return a future with the elements that match the key
fun groupParTuples[k, i](x : k, p : Par[(k, i)], cmp : (k, k) -> int) : Fut[(k, Par[i])]
  val closure = fun (acc : (k, Par[i]), elem : (k, i))
                  val accKey = acc.0
                  val p = acc.1
                  val elemKey = elem.0
                  val elemValue = elem.1
                  val equals = cmp(accKey, elemKey)
                  if equals == 0 then
                    (accKey, p ||| liftv(elemValue))
                  else
                    acc
                  end
                end
  reduce(closure, (x, empty[i]()), p)
end

-- private function
-- finds if an element exists in a ParT of tuples
fun findElemInParTuple[k, t](key : k, p : Par[(k, t)], cmp : (k, k) -> int) : Fut[Maybe[t]]
  val closure = fun (acc : Maybe[t], elem : (k, t))
                  val keyElem = elem.0
                  val keyValue = elem.1
                  if cmp(key, keyElem) == 0 then
                     Just(keyValue)
                  else
                     acc
                  end
                end
  reduce(closure, Nothing, p)
end



{-
Group elements of two ParTs based on their keys and joins them according
to their join function. Elements that have no matching key are ignored.

E.g.

```
  val magnus = new Person("Magnus")
  val terry = new Person("Terry")
  val charlotte = new Person("Charlotte")
  val kiko = new Person("Kiko")

  val barley = new Pet("Barley", terry)
  val boots = new Pet("Boots", terry)
  val whiskers = new Pet("Whiskers", charlotte)
  val daisy = new Pet("Daisy", magnus)

  val people = liftv(magnus) ||| liftv(terry) ||| liftv(charlotte) ||| liftv(kiko)
  val pets = liftv(barley) ||| liftv(boots) ||| liftv(whiskers) ||| liftv(daisy)
  val query = groupJoin(people,
                        pets,
                        fun (p : Person) => p.name,
                        fun (p : Pet) => p.owner.name,
                        fun (per : Person, animals : Par[Pet])
                          liftv((per.name, animals >> (fun (p : Pet) => p.name)))
                        end,
                        fun (s1 : String, s2 : String) => s1.compare(s2))

  for elem <- extract(query) do
    match elem with
      case (name, pets) =>
        println("{}:", name)
        for pet <- extract(pets) do
          println("{}", pet)
        end
      end
    end
  end
```

In the example above, the expected output is:

```
  Terry:
  Barley
  Boots
  Charlotte:
  Whiskers
  Magnus:
  Daisy
```

where the person `Kiko` doesn't have any pet and is therefore ignored.

-}

fun groupJoin[t, i, k, r](pt : Par[t], pi : Par[i], tk : t -> k, ik : i -> k,
                          fn : (t, Par[i]) -> Par[r], cmp : (k, k) -> int) : Par[r]
  -- Pair function with their corresponding ParT
  val pkt = pairWith(pt, tk)
  val pki = pairWith(pi, ik)

  -- Get distinct elements
  val pkid = distinct(pki, fun (v : (k, i), u : (k, i))
                             cmp(v.0, u.0)
                           end)

  join(pkid >> (fun (tuple : (k, i))
                  val key = tuple.0
                  -- group values from the same key under the same ParT
                  val groupedPiKeyToPar = groupParTuples(key, pki, cmp)

                  -- fetch only the values from the tuple
                  val groupedPi = join(liftf(groupedPiKeyToPar ~~> (fun (keyParT: (k, Par[i]))
                                                                      keyParT.1
                                                                    end)))

                  -- find if the key from second ParT (from pkdi) exists
                  -- in the pkt ParT and execute function asynchronously
                  -- only if it does exist.
                  val futtkElem = findElemInParTuple(key, pkt, cmp)
                  val fp = futtkElem ~~> fun (tkElem : Maybe[t])
                                           match tkElem with
                                             case Nothing =>
                                               empty[r]()
                                             end
                                             case Just(tkResultVal) =>
                                               fn(tkResultVal, groupedPi)
                                             end
                                           end
                                         end

                  -- the result is an fp, Fut[Par[r]].
                  -- Lift the future to a ParT and flatten the Par[Par[t]] to Par[r]
                  join(liftf(fp))
                end))
end
