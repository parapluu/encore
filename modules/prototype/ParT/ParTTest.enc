module ParTTest

import ParT

def toUpperAndSmall(s: String): Par String
  liftv(s.to_upper()) || liftv(s)

def printUpper(s: String): void
  print s.to_upper()

class Test
  def producer(): String
    "producer"

class Main
  def testBindAndFilter(): void
    -- this test filters and creates two versions of the filtered items
    let fut = (new Test).producer()
        p = empty<String>() || liftv("Test") || liftv("Quick") ||
            liftv("Encore") || liftf(fut)
        filteredP = filter<String>(\(s: String) ->
                                     s.contains("k") or s.contains("prod"), p)
        bindP = bind<String, String>(toUpperAndSmall, filteredP)
    in for s in extract(bindP) print(s)

  def testReduce(): void
    -- this test concats items in a ParT
    let p = empty<String>() || liftv("Test") || liftv("Quick") || liftv("Encore")
        concatFn = \(x: String, acc: String) -> x.concatenate(acc)
        red = reduce<String, String>(concatFn, "", p)
        result = get(red)
    in print(result)

  def testForeach(): void
    let p = empty<String>() || liftv("Test") || liftv("Quick") || liftv("Encore")
    in foreach<String>(p, printUpper)

  def testAggregate(): void
    let p = liftv("Test") || liftv("Quick") || liftv("Encore")
        result = aggregate<String, int, int>(p, 0,
                                             \(s: String, acc: int) -> acc + 1,
                                             \(x: int) -> x)
                 >> \(total: int) ->
                      "Aggregation result: ".concatenate(string_from_int(total))
    in print((extract(result))[0])

  def main(): void {
    this.testBindAndFilter();
    this.testReduce();
    this.testForeach();
    this.testAggregate();
  }
