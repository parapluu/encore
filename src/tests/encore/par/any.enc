import ParT.ParT

def liftValues(max: int): Par int {
  var p = empty<int>();
  repeat i <- max {
    p = p || liftv(i);
  };
  p
}

def liftFutureValues(max: int): Par int {
  var p = empty<int>();
  repeat i <- max {
    p = p || liftf((new T).set(i));
  };
  p
}

class T {
  def set(x: int): int
    x
}

def lessThanOneHundred(x: int): bool
  x < 100

class Main
  def testValues(): void {
    println("Testing Values:");
    val p = liftValues(100);
    val f = any(p, lessThanOneHundred);
    println("{}", get(f));

    val p2 = liftValues(200);
    val f2 = any(p2, lessThanOneHundred);
    println("{}", get(f2));

    println("{}", get(any(liftv(203), lessThanOneHundred)));
  }

  def testFutures(): void {
    println("Testing Futures:");
    val p = liftFutureValues(100);
    val f = any(p, lessThanOneHundred);
    println("{}", get(f));

    val p2 = liftFutureValues(200);
    val f2 = any(p2, lessThanOneHundred);
    println("{}", get(f2))
  }

  def main(): void {
    this.testValues();
    this.testFutures();
  }
