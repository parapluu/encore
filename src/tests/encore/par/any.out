Testing Values:
true
true
false
Testing Futures:
true
true
